docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md

```diff
@@ -0,0 +1,309 @@
+- 500 Internal Server Error
+  - Generic error that contains further description in the problem detail.
+
+Each error code should have clear guidance in the documentation and API reference for how they should be handled. The problem detail object can provide guidance as well.
+
+## Data fetching
+
+Most resources provide at least one endpoint to fetch related data. Most of those endpoints provide data with near-real time consistency that is queried from exported records, if records for the respective resource are exported by the engine. If resources are not based on exported records, e.g. license data or topology information, the data returned by those endpoints can reflect real time insights or static content.
+
+For most resources, there are search endpoints to query by POST method and a given query request object where applicable. The structure of such query requests always follows the same schema and so does the response, always returning a list of items matching the query criteria.
+
+[Search requests](#search-requests) are forwarded as queries to the datastores that hold the exported records and the query results are returned in the format described in [search responses](#search-responses).
+
+Resources can also support querying subordinate resources. For example, for users and groups, with user search being available at _POST /v2/users/search_, a user's groups can be retrieved using _POST /v2/users/{userKey}/groups/search_. Each resource determines independently if subordinate resources can be accessed this way.
+
+Search endpoints can also be used to directly access entity instances with a unique identifier. As an alternative, resources can also provide GET method endpoints for fetching the data of single instances. In most cases, this is done by a specific key parameter in the URL path, e.g. _GET /v2/users/{userKey}_.
+
+### Search requests
+
+Query requests consist of the components for **filter**, **sort**, and **page**.
+
+The **filter** object defines which fields should match. Only items that match the given fields will be returned. The available fields vary by object and are described in the respective search endpoint. Filtering by a unique identifier is usually available in filtering options. Beyond that, the filter options don’t have to comprise all the returned items’ attributes.
+
+The **sort** object specifies which fields of the object should be sorted and whether they are sorted in ascending (ASC) or descending (DESC) order.
+
+The **page** object details how to slice the result set. An initial search request can omit the page object or define the _limit_. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned _firstItemSortValues_ and _lastItemSortValues_ of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes _searchAfter_ or _searchBefore_.
```

Comment:
This is really helpful, actually...I was going to ask you a question about these fields today anyway. 👍

Reply:
@tmetzke stopping back after attempting to add spec descriptions for these properties, armed with my new understanding -- one other field that isn't mentioned here is `from`, an int32. 

1. Am I correct in thinking this is probably an index to start from? e.g. `from: 11` means start at the 11th item. 
2. If I am correct, is that a zero-based index or 1-based?
3. Do you think we should describe that here, since we covered the other fields?

Reply:
Potentially blocking: 

Since these are fields, I think they should be `code` instead of _italic_. @christinaausley can probably confirm, but that's how I interpret the [code blocks section of the style guide](https://confluence.camunda.com/pages/viewpage.action?spaceKey=HAN&title=Camunda+style+guide).

```suggestion
The **page** object details how to slice the result set. An initial search request can omit the page object or define the `limit`. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned `firstItemSortValues` and `lastItemSortValues` of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes `searchAfter` or `searchBefore`.
```

Reply:
Also blocking:

The [properties in the spec](https://github.com/camunda/camunda/blob/main/zeebe/gateway-protocol/src/main/proto/rest-api.yaml#L4345-L4349) are named `firstSortValues` and `lastSortValues`.

```suggestion
The **page** object details how to slice the result set. An initial search request can omit the page object or define the _limit_. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned _firstSortValues_ and _lastSortValues_ of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes _searchAfter_ or _searchBefore_.
```